### PR TITLE
Fix for character length of secret, validation of name in create secret

### PR DESCRIPTION
**Pull request by CarolynMabbott**
# Changes

Fix issues #566,  #558 and #571 

Changed validation to only allow a secret with less than 253 characters, added validation so secret cant contain . or start / end with a -
Added tests to check that validation works 
Changed error messages on Create Secret so they no longer just say `Required.`

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [ ] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/dashboard/blob/master/CONTRIBUTING.md)
for more details._

### CODE REVIEW

**Comment by a-roberts** [comment]
Optimistically added @AlanGreene as a reviewer, if not I'll take a look at this one tomorrow

File: src/components/SecretsModal/Annotations.js [hunk 1]
```diff
@@ -62,7 +62,7 @@ const Annotations = props => {
         <Add className="addIcon" onClick={handleAdd} />
       </div>
       {invalidFields.find(field => field.includes('annotation-value')) !==
-        undefined && <p className="invalidAnnotation">Required.</p>}
+        undefined && <p className="invalidAnnotation">Server URL Required.</p>}
```
Comment (a-roberts) [inline src/components/SecretsModal/Annotations.js]: I'd rather the case be like that of a normal sentence (this is from the create pipelinerun page) to be consistent, e.g.: `Pipeline cannot be empty` so let's go with "Server URL required"

File: src/components/SecretsModal/BasicAuthFields.js [hunk 2]
```diff
@@ -41,12 +41,12 @@ const BasicAuthFields = props => {
...
         id="username"
         autoComplete="off"
-        placeholder=""
+        placeholder="username"
         value={username}
         labelText="Username:"
         onChange={handleChangeTextInput}
         invalid={invalidFields.indexOf('username') > -1}
-        invalidText="Required."
+        invalidText="Username Required."
```
Comment (a-roberts) [inline src/components/SecretsModal/BasicAuthFields.js]: case

File: src/components/SecretsModal/BasicAuthFields.js [hunk 3]
```diff
@@ -57,7 +57,7 @@ const BasicAuthFields = props => {
         labelText="Password/Token:"
         onChange={handleChangeTextInput}
         invalid={invalidFields.indexOf('password') > -1}
-        invalidText="Required."
+        invalidText="Password or Token Required."
```
Comment (a-roberts) [inline src/components/SecretsModal/BasicAuthFields.js]: case

File: src/components/SecretsModal/BasicAuthFields.js [hunk 4]
```diff
@@ -67,7 +67,7 @@ const BasicAuthFields = props => {
         onChange={handleChangeServiceAccount}
         labelText="Service Account"
         invalid={invalidFields.indexOf('serviceAccount') > -1}
-        invalidText="Required."
+        invalidText="Service Account Required."
```
Comment (a-roberts) [inline src/components/SecretsModal/BasicAuthFields.js]: case

File: src/components/SecretsModal/UniversalFields.js [hunk 5]
```diff
@@ -38,7 +38,7 @@ const UniversalFields = props => {
         labelText="Name:"
         onChange={handleChangeTextInput}
         invalid={invalidFields.indexOf('name') > -1}
-        invalidText="Must be less than 253 characters, contain only lowercase alphanumeric characters, . or -"
+        invalidText="Must not start or end with - and be less than 253 characters, contain only lowercase alphanumeric characters or -"
```
Comment (a-roberts) [inline src/components/SecretsModal/UniversalFields.js]: Really, only lowercase alphanumeric chars? Lowercase?!

File: src/components/SecretsModal/UniversalFields.js [hunk 6]
```diff
@@ -53,7 +53,7 @@ const UniversalFields = props => {
         }
         onChange={handleChangeNamespace}
         invalid={invalidFields.indexOf('namespace') > -1}
-        invalidText="Required."
+        invalidText="Namespace Required."
```
Comment (a-roberts) [inline src/components/SecretsModal/UniversalFields.js]: case

File: src/containers/SecretsModal/SecretsModal.test.js [hunk 7]
```diff
@@ -134,3 +134,314 @@ it('Test SecretsModal click events', () => {
   fireEvent.click(queryByText('Submit'));
   expect(handleSubmit).toHaveBeenCalledTimes(0);
 });
+
+const nameValidationErrorMsgRegExp = /Must not start or end with - and be less than 253 characters, contain only lowercase alphanumeric characters or -/i;
+const namespaceValidationErrorRegExp = /Namespace Required./i;
```
Comment (a-roberts) [inline src/containers/SecretsModal/SecretsModal.test.js]: Case for these

File: src/containers/SecretsModal/SecretsModal.test.js [hunk 8]
```diff
@@ -134,3 +134,314 @@ it('Test SecretsModal click events', () => {
...
+const usernameValidationErrorRegExp = /Username Required./i;
+const passwordValidationErrorRegExp = /Password or Token Required./i;
+const serviceAccountValidationErrorRegExp = /Service Account Required./i;
+const serverurlValidationErrorRegExp = /Server URL Required./i;
+
+const props = {
+  open: true
+};
+
+// describe('Create Secret validation'), () => {
```
Comment (a-roberts) [inline src/containers/SecretsModal/SecretsModal.test.js]: We gonna add it in a describe or nah? No point leaving this commented 😄

File: src/containers/SecretsModal/SecretsModal.test.js [hunk 9]
```diff
@@ -134,3 +134,314 @@ it('Test SecretsModal click events', () => {
...
+  fireEvent.click(queryByText('Submit'));
+  expect(queryByText(nameValidationErrorMsgRegExp)).toBeTruthy();
+  expect(queryByText(namespaceValidationErrorRegExp)).toBeTruthy();
+  expect(queryByText(usernameValidationErrorRegExp)).toBeTruthy();
+  expect(queryByText(passwordValidationErrorRegExp)).toBeTruthy();
+  expect(queryByText(serviceAccountValidationErrorRegExp)).toBeTruthy();
+  expect(queryByText(serverurlValidationErrorRegExp)).toBeTruthy();
+});
+
+it('Create Secret doesnt error when contains - in the middle of the secret', () => {
```
Comment (a-roberts) [inline src/containers/SecretsModal/SecretsModal.test.js]: doesn't* and then make the it be in quotes?

File: src/containers/SecretsModal/SecretsModal.test.js [hunk 10]
```diff
@@ -134,3 +134,314 @@ it('Test SecretsModal click events', () => {
...
+  fireEvent.click(queryByText('Submit'));
+  expect(queryByText(nameValidationErrorMsgRegExp)).toBeFalsy();
+  expect(queryByText(namespaceValidationErrorRegExp)).toBeTruthy();
+  expect(queryByText(usernameValidationErrorRegExp)).toBeTruthy();
+  expect(queryByText(passwordValidationErrorRegExp)).toBeTruthy();
+  expect(queryByText(serviceAccountValidationErrorRegExp)).toBeTruthy();
+  expect(queryByText(serverurlValidationErrorRegExp)).toBeTruthy();
+});
+
+it('Create Secret doesnt error when contains 0', () => {
```
Comment (a-roberts) [inline src/containers/SecretsModal/SecretsModal.test.js]: Same as above

File: src/containers/SecretsModal/SecretsModal.test.js [hunk 11]
```diff
@@ -134,3 +134,314 @@ it('Test SecretsModal click events', () => {
...
+  fireEvent.click(queryByText('Submit'));
+  expect(queryByText(nameValidationErrorMsgRegExp)).toBeFalsy();
+  expect(queryByText(namespaceValidationErrorRegExp)).toBeTruthy();
+  expect(queryByText(usernameValidationErrorRegExp)).toBeTruthy();
+  expect(queryByText(passwordValidationErrorRegExp)).toBeTruthy();
+  expect(queryByText(serviceAccountValidationErrorRegExp)).toBeTruthy();
+  expect(queryByText(serverurlValidationErrorRegExp)).toBeTruthy();
+});
+
+it('Create Secret doesnt error when contains 9', () => {
```
Comment (a-roberts) [inline src/containers/SecretsModal/SecretsModal.test.js]: Same as above

File: src/containers/SecretsModal/SecretsModal.test.js [hunk 12]
```diff
@@ -134,3 +134,314 @@ it('Test SecretsModal click events', () => {
...
+  fireEvent.click(queryByText('Submit'));
+  expect(queryByText(nameValidationErrorMsgRegExp)).toBeFalsy();
+  expect(queryByText(namespaceValidationErrorRegExp)).toBeTruthy();
+  expect(queryByText(usernameValidationErrorRegExp)).toBeTruthy();
+  expect(queryByText(passwordValidationErrorRegExp)).toBeTruthy();
+  expect(queryByText(serviceAccountValidationErrorRegExp)).toBeTruthy();
+  expect(queryByText(serverurlValidationErrorRegExp)).toBeTruthy();
+});
+
+it('Create Secret error when contains 253 characters', () => {
```
Comment (a-roberts) [inline src/containers/SecretsModal/SecretsModal.test.js]: errors*

File: src/containers/SecretsModal/SecretsModal.test.js [hunk 13]
```diff
@@ -134,3 +134,314 @@ it('Test SecretsModal click events', () => {
...
+  fireEvent.click(queryByText('Submit'));
+  expect(queryByText(nameValidationErrorMsgRegExp)).toBeTruthy();
+  expect(queryByText(namespaceValidationErrorRegExp)).toBeTruthy();
+  expect(queryByText(usernameValidationErrorRegExp)).toBeTruthy();
+  expect(queryByText(passwordValidationErrorRegExp)).toBeTruthy();
+  expect(queryByText(serviceAccountValidationErrorRegExp)).toBeTruthy();
+  expect(queryByText(serverurlValidationErrorRegExp)).toBeTruthy();
+});
+
+it('Create Secret doesnt error when contains 252 characters', () => {
```
Comment (a-roberts) [inline src/containers/SecretsModal/SecretsModal.test.js]: same as above

File: src/containers/SecretsModal/SecretsModal.test.js [hunk 14]
```diff
@@ -134,3 +134,314 @@ it('Test SecretsModal click events', () => {
...
+  fireEvent.click(queryByText('Submit'));
+  expect(queryByText(nameValidationErrorMsgRegExp)).toBeFalsy();
+  expect(queryByText(namespaceValidationErrorRegExp)).toBeTruthy();
+  expect(queryByText(usernameValidationErrorRegExp)).toBeTruthy();
+  expect(queryByText(passwordValidationErrorRegExp)).toBeTruthy();
+  expect(queryByText(serviceAccountValidationErrorRegExp)).toBeTruthy();
+  expect(queryByText(serverurlValidationErrorRegExp)).toBeTruthy();
+});
+
+it('Create Secret doesnt error when contains a valid namespace', () => {
```
Comment (a-roberts) [inline src/containers/SecretsModal/SecretsModal.test.js]: same as above

File: src/containers/SecretsModal/SecretsModal.test.js [hunk 15]
```diff
@@ -134,3 +134,314 @@ it('Test SecretsModal click events', () => {
...
+  fireEvent.click(queryByText('Submit'));
+  expect(queryByText(nameValidationErrorMsgRegExp)).toBeFalsy();
+  expect(queryByText(namespaceValidationErrorRegExp)).toBeFalsy();
+  expect(queryByText(usernameValidationErrorRegExp)).toBeTruthy();
+  expect(queryByText(passwordValidationErrorRegExp)).toBeTruthy();
+  expect(queryByText(serviceAccountValidationErrorRegExp)).toBeTruthy();
+  expect(queryByText(serverurlValidationErrorRegExp)).toBeTruthy();
+});
+
+it('Create Secret doesnt error when Docker Registry is selected', () => {
```
Comment (a-roberts) [inline src/containers/SecretsModal/SecretsModal.test.js]: same as above

File: src/containers/SecretsModal/SecretsModal.test.js [hunk 16]
```diff
@@ -134,3 +134,314 @@ it('Test SecretsModal click events', () => {
...
+  fireEvent.click(queryByText('Submit'));
+  expect(queryByText(nameValidationErrorMsgRegExp)).toBeFalsy();
+  expect(queryByText(namespaceValidationErrorRegExp)).toBeFalsy();
+  expect(queryByText(usernameValidationErrorRegExp)).toBeTruthy();
+  expect(queryByText(passwordValidationErrorRegExp)).toBeTruthy();
+  expect(queryByText(serviceAccountValidationErrorRegExp)).toBeTruthy();
+  expect(queryByText(serverurlValidationErrorRegExp)).toBeTruthy();
+});
+
+it('Create Secret doesnt error when a username is entered', () => {
```
Comment (a-roberts) [inline src/containers/SecretsModal/SecretsModal.test.js]: same as above

File: src/containers/SecretsModal/SecretsModal.test.js [hunk 17]
```diff
@@ -134,3 +134,314 @@ it('Test SecretsModal click events', () => {
...
+  fireEvent.click(queryByText('Submit'));
+  expect(queryByText(nameValidationErrorMsgRegExp)).toBeFalsy();
+  expect(queryByText(namespaceValidationErrorRegExp)).toBeFalsy();
+  expect(queryByText(usernameValidationErrorRegExp)).toBeFalsy();
+  expect(queryByText(passwordValidationErrorRegExp)).toBeTruthy();
+  expect(queryByText(serviceAccountValidationErrorRegExp)).toBeTruthy();
+  expect(queryByText(serverurlValidationErrorRegExp)).toBeTruthy();
+});
+
+it('Create Secret doesnt error when a password is entered', () => {
```
Comment (a-roberts) [inline src/containers/SecretsModal/SecretsModal.test.js]: same as above

**Comment by CarolynMabbott** [comment]
> everything looks good for me 👍 the only comment i have is to, if you all agree, move some test cases to the ones inside `components/SecretsModal` folder just to keep `SecretsModal.test` uncluttered.

I havent moved it into components in this PR as it may cause some huge breakages. I have created an issue for it to be moved into components though

File: src/containers/SecretsModal/SecretsModal.js [hunk 18]
```diff
@@ -23,21 +23,32 @@ import { getServiceAccounts, isWebSocketConnected } from '../../reducers';
...
     }
   }
+  if (id === 'name') {
+    if (trimmed.startsWith('-', 0) || trimmed.endsWith('-')) {
+      return false;
+    }
+    if (namespace === undefined) {
+      return false;
+    }
+    if (namespace === 'tekton-pipelines') {
```
Comment (AlanGreene) [inline src/containers/SecretsModal/SecretsModal.js]: this isn't needed as it'll just fall through to the `return true` below anyway

Comment (CarolynMabbott) [reply src/containers/SecretsModal/SecretsModal.js]: My bad, i left that in from testing something worked when helping Megan
Removing it

**Comment by a-roberts** [comment]
^ also fix up your commit messages please @CarolynMabbott, should be a quick rebase job 😄

**Pull request merged**